src/unix/thread.c

```diff
@@ -240,7 +247,8 @@ static void thread_resume(context ctx)
 static void thread_schedule_return(context ctx)
 {
     thread t = (thread)ctx;
-    assert(enqueue_irqsafe(t->scheduling_queue, &t->thread_return));
+    thread_cputime_update(t);   /* so that it is scheduled based on how much CPU time it used */
```

Comment:
Wouldn't the thread be paused at this point, with t->start_time == 0, thus making the call to thread_cputime_update() a no-op here?

Reply:
When a user thread is interrupted by a hardware interrupt, the interrupt handler calls context_schedule_return() without the thread being paused. Also, when the Unix fault handler raises a signal or resolves a page fault without blocking, it calls schedule_thread(), again without the thread being paused.